### ISSUE TITLE
[Perf] Move argmax from host to device for prefill for Falcon7b=B; +10–15% throughput

### ISSUE THREAD

**Issue by aliaksei-sala** [tenstorrent/tt-metal#30883]
### Is your feature request related to a problem? Please describe

If we move the argmax (next-token selection) computation from the host CPU to the device, eliminating a host-device round-trip and CPU-side post-processing during prefill, this improves prefill inference throughput by ~10–15% in our benchmarks.

# Motivation

During prefill, we were transferring full logits back to the host solely to compute argmax. That introduced:
- Extra PCIe bandwidth and latency
- CPU work on a latency-sensitive path
- Synchronization stalls between device and host

Computing argmax on-device lets us return only token IDs (or a tiny top-k), shrinking transfers and removing a CPU bottleneck.

# Impact

Device: [n150 (Wormhole)](https://tenstorrent.com/hardware/wormhole)
Performance was measured with the following command:
`pytest --disable-warnings -q -s --input-method=json --input-path='models/demos/falcon7b_common/demo/input_data.json' models/demos/wormhole/falcon7b/demo_wormhole.py::test_demo -k "perf_mode_2048_stochastic and not verify"`

Verified with : 
`pytest -q /home/aliaksei/wp/tt/tt-metal/models/demos/wormhole/falcon7b/demo_wormhole.py -k "default_mode_1024_greedy_verify"`

- Throughput (prefill): +10–15%
- Host↔Device traffic: reduced (logits → token IDs/top-k)
- CPU utilization: reduced during prefill

Loading weights, compile times were inconsistent across runs

| Metric                                              | Before   | After    | Δ (After-Before) | % Change | Trend |
| --------------------------------------------------- | -------- | -------- | ---------------- | -------- | ----- |
| pre processing                                      | 0.275    | 0.273    | -0.002           | +0.64%   | 🟡→     |
| loading weights                                     | 3.022    | 2.075    | -0.946           | +31.31%  | 🟢↓     |
| conversion to TT and moving weights to device       | 7.939    | 6.910    | -1.029           | +12.96%  | 🟢↓     |
| initializing KV cache                               | 0.186    | 0.206    | +0.020           | -10.62%  | 🔴↑     |
| prefill compile time (single layer)                 | 7.344    | 7.094    | -0.250           | +3.40%   | 🟢↓     |
| decode compile time (single layer)                  | 7.660    | 7.519    | -0.140           | +1.83%   | 🟢↓     |
| total compile time (single layer)                   | 15.003   | 14.613   | -0.391           | +2.60%   | 🟢↓     |
| prefill inference time                              | 8.611    | 7.506    | -1.105           | +12.83%  | 🟢↓     |
| decode inference time                               | 1.297    | 1.292    | -0.005           | +0.41%   | 🟢↓     |
| total inference time                                | 9.908    | 8.798    | -1.110           | +11.21%  | 🟢↓     |
| time to first token (prefill, 1st user)             | 0.861    | 0.751    | -0.110           | +12.83%  | 🟢↓     |
| inference throughput prefill (users/s)              | 1.16     | 1.33     | +0.17            | +14.72%  | 🟢↑     |
| inference throughput, prefill seq_len=2048 (tok/s) | 2,378.41 | 2,728.55 | +350.14          | +14.72%  | 🟢↑     |
| inference throughput decode (tok/s)                 | 493.39   | 495.40   | +2.01            | +0.41%   | 🟢↑     |
| inference throughput decode (per user)              | 15.42    | 15.48    | +0.06            | +0.41%   | 🟢↑     |
| inference throughput prefill+decode (per user)      | 1.08     | 1.23     | +0.15            | +13.59%  | 🟢↑     |
| Total demo duration                                 | 41.870   | 38.300   | -3.570           | +8.53%   | 🟢↓     |
| Total number of tokens generated in decode          | 640      | 640      | +0.00            | +0.00%   | 🟡→     |

# Tracy 
Before optimization:
<img width="992" height="455" alt="Image" src="https://github.com/user-attachments/assets/27d9f6a0-3699-4ec0-85da-4c99cefbeba1" />

After optimization:
<img width="1108" height="394" alt="Image" src="https://github.com/user-attachments/assets/76e1e67b-7076-44f8-9504-6fd39e0cd0b2" />

### Describe the solution you'd like

Move the argmax computation from the host CPU to the device during prefill stage for Falcon7B
